src/test/java/com/google/cloud/spanner/pgadapter/python/PythonTransactionTests.java

```diff
@@ -0,0 +1,1095 @@
+    statements.add("query");
+    statements.add(sql);
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql), createResultSet(1, "abcd")));
+
+    String expectedOutput = "the 'deferrable' setting is only available from PostgreSQL 9.1\n";
+    String actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+
+    Assert.assertEquals(expectedOutput, actualOutput);
+
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(CommitRequest.class));
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+  }
+
+  @Test
+  public void testDeferrableSessionInTransactions() throws IOException, InterruptedException {
+    List<String> statements = new ArrayList<>();
+    String sql = "Select * from some_table";
+
+    // It will throw an error because deferrable is not supported
+    statements.add("transaction");
+    statements.add("set session deferrable True");
+
+    // statement won't be executed because the above setting will throw error
+    statements.add("query");
+    statements.add(sql);
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql), createResultSet(1, "abcd")));
+
+    String expectedOutput = "the 'deferrable' setting is only available from PostgreSQL 9.1\n";
+    String actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+
+    Assert.assertEquals(expectedOutput, actualOutput);
+
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(CommitRequest.class));
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+  }
+
+  @Test
+  public void testReadOnlyInTransactions() throws IOException, InterruptedException {
+    // tests readonly settings using connection.readonly variable
+    List<String> statements = new ArrayList<>();
+    String sql1 = "Select * from some_table";
+    String sql2 = "insert into some_table(col1, col2) values(value1, value2)";
+    String sql3 = "Select * from some_table3";
+    String sql4 = "insert into some_table3(col1, col2) values(value1, value2)";
+
+    // query will be executed as expected
+    statements.add("query");
+    statements.add(sql1);
+
+    // update will be executed as expected
+    statements.add("update");
+    statements.add(sql2);
+
+    // update will be executed as expected
+    statements.add("update");
+    statements.add(sql4);
+
+    // 1 commit request will be sent
+    statements.add("transaction");
+    statements.add("commit");
+
+    statements.add("transaction");
+    statements.add("set readonly True");
+
+    // query will be executed as expected
+    statements.add("query");
+    statements.add(sql3);
+
+    // update will throw an error because the readonly is activated
+    statements.add("update");
+    statements.add(sql2);
```

Comment:
It's good that we test this, but we should also test a successful read-only transaction that is committed by the application. Committing a read-only transaction in psycopg2 is allowed (and required to release all resources associated with the transaction). It should however not lead to a `CommitRequest` being sent to Cloud Spanner, as Cloud Spanner does not require read-only transactions to be committed or rolled back.

Reply:
I've added this test also in the same test case.